### DISCUSSION TITLE
Better JSVM integration

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3913] [Q&A]
It would be nice if there was a function signature for binding packages to JSVM.  Only certain things need to be exposed, and without this a lot of packages will have a lot of crappy boiletplate outside of it.  I'd also like the JSVM binding to be able to name the exposed `$` or `$$` variables from within the binding.  These bindings could return a map that is iterated upon to add to the JSVM.

So for example, I want my new camera route to be addable to the JSVM with just:

```golang
jsvm.Add(camera.JSVM)
```

So if it's a special app-specific tech then it would be done similarly to how some people use http controllers.

```golang
jsvm.Add(myAppJSVM(myAppStruct))
```

Where I'm going with this is people being able to go with what is basically a custom build of Pocketbase that pretty much just adds some JSVM bindings, and isn't meant to be used in Golang.  I'll be doing a combination of this for my Raspberry Pi.

The JSVM environment is far superior to Golang-only for slow computers.  It could also be a skill thing, or just wanting to be able to live-hack on a system like how people can with a PHP site.

About the top-level binding of `$`, I'm suggesting that you reserve that for the build-ins.  So custom could use `$$`.  It will be more obvious for people when they're using extensions, and you won't have to worry about collisions.

**Comment by dschissler** [comment]
The next cool level is to make a build time feature of pocketbase that allows it to read a package that contains the JSVM bindings.  This would allow very easy integration of JSVM tech.

Perhaps default Pocketbase could compile this mostly empty package that wouldn't do much.  For someone to add JSVM bindings they'd only need to add the `import`, then add the line in the function like `jsvm.Add(camera.JSVM)`.  Then `go mod tidy`.  Then their build would have custom stuff.

If you did this then I could provide line 10 lines of Raspberry Pi relevant integrations for someone to paste into their JSVM init function.  Then they're able to toggle pins, etc, and without using Python.

**Comment by ganigeorgiev** [comment]
I have it in my todo but I didn't want to block further the v0.20.0 release as it already included too many other things.

In general the idea is that eventually we'll have an optional bind function as part of `jsvm.Config` struct that will accept a `vm *goja.Runtime` as argument. There are no plans for now to reserve or restrict the keys of the custom bindings as it is possible that developers may want to load a custom goja native module (eg. `fetch`) or overwrite/patch an already existing binding.

**Reply by dschissler** [reply]
That's fine.  I'm currently prefixing my (fake app store) bindings with underscore.  It reads well.
